### ISSUE TITLE
Placing knots, on more than two lineages

### ISSUE THREAD

**Issue by HectorRDB** [statOmics/tradeSeq#1]
When fitting the Gam models with lineages, the end points of lineages replace the value of the closest knots (as previously defined by quantiles)
However, this does not work if several end times are close to the same knot (run into error "Can't get all knots to equal endpoints of trajectories").
I see two options: 
- Add them all as knots (then, we do have more knots here so we may bias the isIdenticalTest towards those times)
- Add the mean as knot. Then, the knot is not really the end time but it is still quite close so it should be good. However, this impacts more tests (startVsEndTest, diffEndTest)
Ideas?

**Comment by zhangjiaobxy** [comment]
Hi,
I tried tradeSeq and also got the same error (see below). Any updates for this issue? Thanks.
> gamList <- fitGAM(counts = counts,
+                   pseudotime = slingPseudotime(crv, na = FALSE),
+                   cellWeights = slingCurveWeights(crv),
+                   verbose = FALSE)
Error in fitGAM(counts = counts, pseudotime = slingPseudotime(crv, na = FALSE),  : 
  Can't get all knots to equal endpoints of trajectories

**Comment by HectorRDB** [comment]
Hi,
You can solve temporarily this issue by adding more knots, with the `nknots` option of __fitGam__. To get some ideas, endpoints of each lineage should be separated by a knot, and knots are spaced on equal quantiles. You can use this info to decide on the appropriate number of knots. 
I am still debating how to address this case in the general setting.

**Comment by zhangjiaobxy** [comment]
Thanks @HectorRDB !
Yes, it works when adding more knots!
BTW, I am not sure what is the parameter `nPoints` in the `patternTest ` fuction. Is `nPoints` means the number of cells? Thanks!

**Comment by koenvandenberge** [comment]
Should we avoid this error by setting equally spaced knots (with a warning) as we did [here](https://github.com/statOmics/tradeSeq/blob/master/R/core.R#L126-L134)?

**Comment by HectorRDB** [comment]
@koenvandenberge I don't think it should be an error anyway? The tests already assume that the endpoints are not (necessarily) knots. We can replace the error by a warning and leave it there?

**Comment by HectorRDB** [comment]
> Thanks @HectorRDB !
> Yes, it works when adding more knots!
> BTW, I am not sure what is the parameter `nPoints` in the `patternTest` fuction. Does `nPoints` indicate the number of cells? Thanks!

@zhangjiaobxy the nPoints argument refers to the number of extrapolated (pseudo)time points used to contrast the lineages. I refer you to our manuscript (https://biorxiv.org/content/10.1101/623397v1) for more details, methods patternTest.

**Comment by koenvandenberge** [comment]
@HectorRDB I agree, this should now be resolved in the latest commit 7bab5d7

**Comment by HectorRDB** [comment]
I think we misunderstood each other. Your commit 7bab5d7 does replace the error by a warning, which we agree makes more sense. But I was also saying that we can keep quantile-spaced knots in that case

**Comment by koenvandenberge** [comment]
Thanks for pointing this out.
I was testing a few things whether this could not lead us to pathological cases, and will revert back to quantile-spaced knots as you suggest.

While testing I also found that if one of the lineages ends close to the longest lineage it can occur that the last knot point is overwritten, and this will thus be lower than the maximum pseudotime value. I have now tried to enforce the first and last knot point in 43f4384.
